pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java

```diff
@@ -0,0 +1,282 @@
+    // possible between the servers.
+    //
+    // Each Kafka partition contains a fraction of the data, so we need to make sure that we query all partitions.
+    // Because in certain unlikely degenerate scenarios, we can consume overlapping data until segments are flushed (at
+    // which point the overlapping data is discarded during the reconciliation process with the controller), we need to
+    // ensure that the query that is sent has only one partition in CONSUMING state in order to avoid duplicate records.
+    //
+    // Because we also want to want to spread the load as equally as possible between servers, we use a weighted random
+    // replica selection that favors picking replicas with fewer segments assigned to them, thus having an approximately
+    // equal distribution of load between servers.
+    //
+    // For example, given three replicas with 1, 2 and 3 segments assigned to each, the replica with one segment should
+    // have a weight of 2, which is the maximum segment count minus the segment count for that replica. Thus, each
+    // replica other than the replica(s) with the maximum segment count should have a chance of getting a segment
+    // assigned to it. This corresponds to alternative three below:
+    //
+    // Alternative 1 (weight is sum of segment counts - segment count in that replica):
+    // (6 - 1) = 5 -> P(0.4166)
+    // (6 - 2) = 4 -> P(0.3333)
+    // (6 - 3) = 3 -> P(0.2500)
+    //
+    // Alternative 2 (weight is max of segment counts - segment count in that replica + 1):
+    // (3 - 1) + 1 = 3 -> P(0.5000)
+    // (3 - 2) + 1 = 2 -> P(0.3333)
+    // (3 - 3) + 1 = 1 -> P(0.1666)
+    //
+    // Alternative 3 (weight is max of segment counts - segment count in that replica):
+    // (3 - 1) = 2 -> P(0.6666)
+    // (3 - 2) = 1 -> P(0.3333)
+    // (3 - 3) = 0 -> P(0.0000)
+    //
+    // Of those three weighting alternatives, the third one has the smallest standard deviation of the number of
+    // segments assigned per replica, so it corresponds to the weighting strategy used for segment assignment. Empirical
+    // testing shows that for 20 segments and three replicas, the standard deviation of each alternative is respectively
+    // 2.112, 1.496 and 0.853.
+    //
+    // This algorithm works as follows:
+    // 1. Gather all segments and group them by Kafka partition, sorted by sequence number
+    // 2. Ensure that for each partition, we have at most one partition in consuming state
+    // 3. Sort all the segments to be used during assignment in ascending order of replicas
+    // 4. For each segment to be used during assignment, pick a random replica, weighted by the number of segments
+    //    assigned to each replica.
+
+    // 1. Gather all segments and group them by Kafka partition, sorted by sequence number
+    Map<String, SortedSet<String>> sortedSegmentsByKafkaPartition = new HashMap<String, SortedSet<String>>();
```

Comment:
Also, it helped on the controller side to add a segment holder class that implements Comparable, and sorts segment names by kafka partition and then by offset (we can change offset to time in the final implementation). Perhaps that will be useful here ? (check the validation branch). If so, please introduce a class that holds the lowlevel segment name and extracts components out of it.